### ISSUE TITLE
Smoothing Methods like Jacobi & Gauss-Seidel

### ISSUE THREAD

**Issue by Ceyron** [patrick-kidger/lineax#116]
Hi Patrick & Jason,

Thanks for this wonderful library. I was curious if you are interested in adding smoothing methods based on matrix decomposition, such as [the Jacobi method](https://en.wikipedia.org/wiki/Jacobi_method) or [Gauss-Seidel](https://en.wikipedia.org/wiki/Gauss%E2%80%93Seidel_method) (potentially with the option for relaxation).

Certainly, these solvers a way less efficient than the Krylov ones and need matrix materialization (the state of sparse matrices in JAX is not mature enough?!?), but they could complete the suite of solvers. They might also serve as a nice educational example.

**Comment by vboussange** [comment]
Hey @Ceyron, this relates to https://github.com/patrick-kidger/lineax/issues/113 and https://github.com/patrick-kidger/lineax/issues/8 . Those methods would be very useful in multigrid methods that could be used for preconditioners to feed into e.g. GMRES. I am currently running a project that needs those. It could be nice to coordinate; I suggested to implement an example with preconditioning for solving a 2D Poisson system in https://github.com/patrick-kidger/lineax/issues/8, although I have not advanced on that yet.

**Comment by Ceyron** [comment]
> I'd be happy to take a PR on these :)

Excellent. Then, we could start working on this in the coming weeks.

@vboussange: Do you already have some code for Jacobi or Gauss-Seidel? Otherwise, I would start looking into subclassing a new iterative solver since I need it for my research anyway.

**Comment by vboussange** [comment]
@Ceyron Nope, I have not started digging into that - great that you can take the lead! I had in mind to get inspired from AMG.jl, specifically [here](https://github.com/JuliaLinearAlgebra/AlgebraicMultigrid.jl/blob/84e4a6a2fb0e3d43ad64b1945f0d199feac356c4/src/smoother.jl) and [there]( https://github.com/JuliaLinearAlgebra/AlgebraicMultigrid.jl/blob/84e4a6a2fb0e3d43ad64b1945f0d199feac356c4/src/multilevel.jl). Let me know if I can help!

**Comment by Ceyron** [comment]
> @Ceyron Nope, I have not started digging into that - great that you can take the lead! I had in mind to get inspired from AMG.jl, specifically [here](https://github.com/JuliaLinearAlgebra/AlgebraicMultigrid.jl/blob/84e4a6a2fb0e3d43ad64b1945f0d199feac356c4/src/smoother.jl) and [there](https://github.com/JuliaLinearAlgebra/AlgebraicMultigrid.jl/blob/84e4a6a2fb0e3d43ad64b1945f0d199feac356c4/src/multilevel.jl). Let me know if I can help!

No problem :+1:  Cool, then I'd start fresh with Jacobi & GaussSeidel. I want to create a first draft and afterward loop you in for some feedback.
